File: whatsapp_gateway/meta_api/signature.py
"""Facebook/Meta webhook signature verification."""

import hashlib
import hmac


def verify_facebook_signature(
    app_secret: str, payload: bytes, sig256: str | None, sig1: str | None
) -> bool:
    """
    Verify the signature from Meta webhook requests.

    Args:
        app_secret: your Meta app secret (string)
        payload: raw request body as bytes
        sig256: value of X-Hub-Signature-256 header (e.g., 'sha256=...')
        sig1: value of X-Hub-Signature header (e.g., 'sha1=...') - legacy fallback

    Returns:
        True if signature is valid, False otherwise
    """
    # Prefer SHA-256 if provided
    if sig256:
        expected = (
            "sha256=" + hmac.new(app_secret.encode("utf-8"), payload, hashlib.sha256).hexdigest()
        )
        return hmac.compare_digest(expected, sig256.strip())

    # Fallback to SHA-1 if only that header is present
    if sig1:
        expected = "sha1=" + hmac.new(app_secret.encode("utf-8"), payload, hashlib.sha1).hexdigest()
        return hmac.compare_digest(expected, sig1.strip())

    return False

File: src/services/meta-api/signature.ts
/**
 * Meta/Facebook webhook signature verification.
 *
 * Uses Web Crypto API (available in Cloudflare Workers) for HMAC verification.
 */

import { arrayBufferToHex, constantTimeCompare } from '../../utils/crypto';

/**
 * Verify the signature from Meta webhook requests.
 *
 * @param body - Raw request body as string
 * @param signature - Value of X-Hub-Signature-256 header (e.g., 'sha256=...')
 * @param secret - Meta app secret
 * @returns Promise<boolean> - True if signature is valid
 */
export async function verifySignature(
  body: string,
  signature: string | null,
  secret: string
): Promise<boolean> {
  if (!signature) {
    return false;
  }

  const encoder = new TextEncoder();

  // Import the secret as a CryptoKey
  const key = await crypto.subtle.importKey(
    'raw',
    encoder.encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );

  // Compute the HMAC signature
  const signatureBuffer = await crypto.subtle.sign('HMAC', key, encoder.encode(body));

  // Convert to hex string with sha256= prefix
  const expected = 'sha256=' + arrayBufferToHex(signatureBuffer);

  // Constant-time comparison to prevent timing attacks
  return constantTimeCompare(expected, signature.trim());
}

/**
 * Verify signature with fallback to SHA-1 for legacy support.
 *
 * @param body - Raw request body as string
 * @param sig256 - Value of X-Hub-Signature-256 header
 * @param sig1 - Value of X-Hub-Signature header (legacy SHA-1)
 * @param secret - Meta app secret
 * @returns Promise<boolean> - True if signature is valid
 */
export async function verifyFacebookSignature(
  body: string,
  sig256: string | null,
  sig1: string | null,
  secret: string
): Promise<boolean> {
  // Prefer SHA-256 if provided
  if (sig256) {
    return verifySignature(body, sig256, secret);
  }

  // Fallback to SHA-1 (legacy)
  if (sig1) {
    return verifySha1Signature(body, sig1, secret);
  }

  return false;
}

async function verifySha1Signature(
  body: string,
  signature: string,
  secret: string
): Promise<boolean> {
  const encoder = new TextEncoder();

  const key = await crypto.subtle.importKey(
    'raw',
    encoder.encode(secret),
    { name: 'HMAC', hash: 'SHA-1' },
    false,
    ['sign']
  );

  const signatureBuffer = await crypto.subtle.sign('HMAC', key, encoder.encode(body));
  const expected = 'sha1=' + arrayBufferToHex(signatureBuffer);

  return constantTimeCompare(expected, signature.trim());
}
